### PR TITLE
Min version cut versions

### PR DESCRIPTION
**Pull request by phadej**
The change motivated by https://github.com/haskell/pvp/issues/4

ping @dcoutts @hvr @bergmark

### CODE REVIEW

**Comment by ezyang** [comment]
This patch would also need to get merged with GHC's MIN_VERSION support.

**Comment by phadej** [comment]
@ezyang I can make a arc patch, after others approve this patch.

**Comment by phadej** [comment]
@ezyang why GHC definitions come after `cabal_macros.h`? It seems to break GHC-8.0 build

**Comment by ezyang** [comment]
Hmm, well, we can solve this problem by having Cabal pass `-fno-version-macros` to suppress GHC's version macros. Ordering doesn't really matter; unless we use ifdef style definitions, redefinition is always a failure.

**Comment by phadej** [comment]
@ezyang, yeah. That's why I added `#ifndef`s in the last commit, hoped that it would fix the build.

**Comment by hvr** [comment]
@phadej can you provide a bit more description/explaination what this patch changes from the user perspective?

**Comment by phadej** [comment]
@hvr, if there is `package-1` and `package-1.0`, then with `MIN_VERSION_package(1,0,0)` one would match only `package-1.0`.

So this is breaking change :(

**Comment by hvr** [comment]
@phadej that's very bad then, we have lots of packages, like @ekmett's which tend to drop redundant trailing 0s (c.f. `nats` or `lens`). So this is IMO a show-stopper.

**Comment by phadej** [comment]
```c
/* tool hscolour-1.22 */
#ifndef TOOL_VERSION_hscolour
#define TOOL_VERSION_hscolour "1.22"
#endif
#ifndef MIN_TOOL_VERSION_hscolour
#define MIN_TOOL_VERSION_hscolour(major1,major2,minor) \
  ((major1) <  1 || \
  (major1) == 1 && (major2) <  22 || \
  (major1) == 1 && (major2) == 22 && (minor) <= 0)
#endif
#ifndef MIN2_TOOL_VERSION_hscolour
#define MIN2_TOOL_VERSION_hscolour(major1,major2,minor) \
  ((major1) <  1 || \
  (major1) == 1 && (major2) <  22 || \
  (major1) == 1 && (major2) == 22 && (minor) <= -1)
#endif
```

so with

```c
#if MIN_TOOL_VERSION_hscolour(1,22,0)
"foobar"
#else
"------"
#endif

#if MIN2_TOOL_VERSION_hscolour(1,22,0)
"foobar"
#else
"------"
#endif

#if MIN2_TOOL_VERSION_hscolour(1,22,-1)
"foobar"
#else
"------"
#endif
```

the output is roughly:

```
"foobar"
"------"
"foobar"
```

**Comment by ekmett** [comment]
I expect this patch would break the vast majority of my code.

**Comment by phadej** [comment]
@ekmett @hvr I left MIN_VERSION version as it was, MIN2_VERSION is one with new behaviour

**Comment by 23Skidoo** [comment]
> `MIN2_VERSION`. 

Can we pick up a better name, e.g. `MIN_VERSION_EXACT_foo`? The new macro also needs to be documented.

File: Cabal/doc/developing-packages.markdown [hunk 1]
```diff
@@ -2153,6 +2153,10 @@ Since version 1.20, there is also the `MIN_TOOL_VERSION_`_`tool`_ family of
 macros for conditioning on the version of build tools used to build the program
 (e.g. `hsc2hs`).
 
+Since version 1.26, there is also `MIN_VERSION_EXACT_`_`package`_ and
+`MIN_TOOL_VERSION_EXACT_`_`tool`_ version, which let you differntiate between
```
Comment (ezyang) [inline Cabal/doc/developing-packages.markdown]: typo

**Comment by ezyang** [comment]
Sorry, the documentation is still insufficient. So when am I supposed to use the exact macro? Am I ONLY ever supposed to use the exact macro (and I'm wrong if I use the old-style macro?) "If there is package-1 and package-1.0, then with MIN_VERSION_package(1,0,0) one would match only package-1.0" is a very useful comment, let's see it in the docs.

Let me even take a step back and ask: is there some de facto interpretation of the PVP which conflicts with the letter of the PVP? Perhaps on the relation of 1.0 and 1? If so, maybe the letter of the PVP should change. I've always thought that 1.0 and 1.0.0 are equivalent; if this were not the case, that's surprising!

**Comment by ekmett** [comment]
Rather than enshrining the existing behavior, perhaps we should just fix it. 

There are maybe a half dozen packages that have both v1 and v1.0 versions, and _nobody cares_. We could fix the version check to effectively zero pad, delete the conflicting versions from hackage, make it so it doesn't accept new instances of this problem and not be stuck explaining this for decades to come.

We're struggling to find a path of least breakage for a misfeature.

**Comment by 23Skidoo** [comment]
Related: #1368.

**Comment by phadej** [comment]
> Rather than enshrining the existing behavior, perhaps we should just fix it.

:+1: as well. I probably should send a proposal to `libraries` list about this? Will do ASAP.

**Comment by ezyang** [comment]
A good test would be query Hackage's index and find out how many packages have this kind of overlap.

In #1368 the discussion got derailed because then we wanted to fix the Haskell Ord instance to also treat these as equal, but then that would imply changing Data.Version, and then someone, somewhere was relying on the old behavior and will complain when it changes. I guess maybe the right thing to do is for Cabal to define its own version type. Then we could get rid of tags too (he he).

**Comment by phadej** [comment]
There are **none** versions on `Hackage` with tags (or my [script](https://gist.github.com/phThere are **none** versions on `Hackage` with tags (or my [script](https://gist.github.com/phadej/5d5ac2be247fb2c81b521eef63c46402) is broken)

Yet: 41 packages with trailing zeroes used as minor or patch level bump: 
- `DSA`:
    - `1 -> 1.0` no change: http://hdiff.luite.com/cgit/DSA/diff?id=1.0&id2=1
    - `1.0 -> 1.0.0` new instance http://hdiff.luite.com/cgit/DSA/diff?id=1.0.0&id2=1.0 
- `Win32` `2.1 -> 2.2.0.0` (pre GHC-7 time) `http://hdiff.luite.com/cgit/Win32/diff?id=2.1.0.0&id2=2.1` added `System.Win32.Security`
- `RSA` `2.0 -> 2.0.0` introduces non-compilation breakage on GHC7.0 and GHC7.2 atm: http://matrix.hackage.haskell.org/package/RSA#GHC-7.0/RSA-2.0.0 (maybe fixable with correct bounds)
- `pipes-cereal`: no difference http://hdiff.luite.com/cgit/pipes-cereal/diff?id=0.1.0.0&id2=0.1.0
- `extensible-effects`: revision would been enough (probably wasn't available back then) http://hdiff.luite.com/cgit/extensible-effects/diff?id=1.7.1.0&id2=1.7.1
- `lzma-conduit`: patch level change in tests: http://hdiff.luite.com/cgit/lzma-conduit/diff?id=1.1.3.0&id2=1.1.3
- `hoodle`:
    - `0.4 -> 0.4.0`: no change: http://hdiff.luite.com/cgit/hoodle/diff?id=0.4.0&id2=0.4
    - `0.2 -> 0.2.0`: metadata change: http://hdiff.luite.com/cgit/hoodle/diff?id=0.2.0&id2=0.2
    
```
DSA: 1 1.0 1.0.0 1.0.1
Dangerous: 0.1 0.1.1 0.1.2 0.2 0.2.1 0.2.2 0.2.3 0.3 0.3.0 0.3.1 0.3.2
RSA: 1.0.0 1.0.1 1.0.2 1.0.3 1.0.4 1.0.5 1.0.6 1.0.6.1 1.0.6.2 1.0.6.3 1.2.0.0 1.2.1.0 1.2.2.0 2 2.0 2.0.0 2.1.0 2.1.0.1 2.1.0.2 2.1.0.3 2.2.0
Win32: 2.1 2.1.0.0 2.2.0.0 2.2.0.1 2.2.0.2 2.2.1.0 2.2.2.0 2.3.0.0 2.3.0.1 2.3.0.2 2.3.1.0 2.3.1.1
Workflow: 0.1 0.2 0.3 0.5.5 0.5.6 0.5.8 0.5.8.0 0.5.8.1 0.5.8.2 0.6.0.0 0.7.0.0 0.7.0.1 0.7.0.2 0.7.0.3 0.7.0.4 0.7.0.5 0.7.0.6 0.7.0.7 0.8.0.0 0.8.0.1 0.8.0.2 0.8.0.3 0.8.0.4 0.8.0.5 0.8.0.6 0.8.0.7 0.8.0.8 0.8.0.9 0.8.1 0.8.2 0.8.3
binary-derive: 0.1 0.1.0
coroutine-object: 0.1 0.1.0 0.2 0.2.0 0.2.0.0 0.3 0.3.0
data-category: 0.0.1 0.0.2 0.0.3 0.0.3.1 0.1.0 0.2.0 0.3.0 0.3.0.1 0.3.0.2 0.3.1 0.3.1.1 0.4 0.4.1 0.5.0 0.5.1 0.5.1.0 0.5.1.1 0.6.0 0.6.1
diff3: 0.1.0.0 0.1.0.1 0.1.2 0.2 0.2.0 0.2.0.1 0.2.0.2 0.2.0.3
extensible-effects: 1.0 1.1.0 1.2.0 1.2.1 1.4.1 1.5.0 1.6.0 1.7.1 1.7.1.0 1.7.1.1 1.7.1.2 1.7.2.1 1.8.0.0 1.8.1.0 1.9.0.0 1.9.0.1 1.9.1.0 1.9.2.2 1.10.0.1 1.11.0.0 1.11.0.1 1.11.0.2 1.11.0.3 1.11.0.4
fficxx: 0.1 0.1.0 0.2 0.2.1
homplexity: 0.2 0.2.0.0 0.3.0.0 0.4.0.0 0.4.1.0 0.4.2.0 0.4.3.0 0.4.3.1 0.4.3.2 0.4.3.3
hoodle: 0.1 0.1.0 0.1.0.0 0.1.0.1 0.1.1 0.1.1.0 0.1.1.1 0.2 0.2.0 0.2.1 0.2.2 0.2.2.0 0.2.2.1 0.3 0.4 0.4.0
hoodle-builder: 0.1 0.1.0 0.1.1 0.2 0.2.1 0.2.2 0.2.2.0 0.3 0.3.0
hoodle-core: 0.7 0.7.0 0.7.0.0 0.7.1 0.8 0.8.0 0.9 0.9.0 0.9.0.0 0.10 0.11 0.12 0.13 0.13.0 0.13.0.0 0.14 0.15 0.15.0
hoodle-parser: 0.1 0.1.0 0.1.1 0.2 0.2.1 0.2.2 0.3 0.3.0
hoodle-publish: 0.1 0.2 0.2.0
hoodle-render: 0.1 0.1.0 0.2 0.2.0 0.2.1 0.3 0.3.1 0.3.2 0.4 0.5 0.5.0
hoodle-types: 0.1 0.1.0 0.1.1 0.2 0.2.1 0.2.2 0.3 0.3.0
hxournal: 0.5 0.5.0.0 0.5.1 0.6 0.6.0 0.6.0.0 0.6.0.1 0.6.1 0.6.1.0 0.6.2 0.6.2.0 0.6.2.1 0.6.3 0.6.4 0.6.4.0 0.6.4.1 0.6.4.2 0.6.4.3 0.6.5 0.6.5.0 0.6.6 0.6.6.1
io-capture: 0.1 0.1.0.0 0.2 0.3 1.0.0
level-monad: 0.1 0.2 0.2.1 0.2.2 0.2.2.0 0.2.2.1 0.3 0.3.1 0.4 0.4.1
lojban: 0.0 0.0.0 0.1 0.2 0.3
lzma-conduit: 0.1.2 0.1.2.1 0.1.2.2 0.4.0 0.4.1 0.4.2 0.5.2 0.5.2.1 0.5.2.2 0.5.2.3 1.0.8 1.1.0 1.1.1 1.1.2 1.1.3 1.1.3.0
monad-parallel: 0.5 0.5.1 0.5.2 0.7 0.7.1 0.7.1.1 0.7.1.2 0.7.1.3 0.7.1.4 0.7.2 0.7.2.0 0.7.2.1 0.7.2.2
open-union: 0.1.0 0.1.0.0 0.1.0.1
pipes-cereal: 0.1.0 0.1.0.0
poppler: 0.11.0 0.11.1 0.12.0 0.12.1 0.12.2 0.12.2.1 0.12.2.2 0.12.3 0.13 0.13.1 0.14 0.14.0
scgi: 0.1 0.1.0 0.2 0.2.1 0.3 0.3.1 0.3.2 0.3.3
spelling-suggest: 0.5.0 0.5.0.1 0.5.1 0.5.1.0 0.5.2.0 0.5.2.1
storablevector: 0.1 0.1.0 0.1.1 0.1.2 0.1.2.2 0.2 0.2.1 0.2.2 0.2.3 0.2.3.1 0.2.4 0.2.5 0.2.6 0.2.7 0.2.7.1 0.2.7.2 0.2.7.3 0.2.8 0.2.8.1 0.2.8.2 0.2.8.3 0.2.9 0.2.9.1 0.2.10
stream-monad: 0.1 0.1.1 0.1.1.0 0.2 0.3 0.4 0.4.0.1 0.4.0.2
t3-game: 0.1.0 0.1.0.0 1.1.0 2.0.0
ttrie: 0.1 0.1.0.0 0.1.1 0.1.2 0.1.2.1
typography-geometry: 1.0 1.0.0 1.0.0.1
wxcore: 0.10.1 0.10.2 0.10.3 0.10.4 0.10.5 0.10.6 0.10.7 0.10.8 0.10.9 0.10.10 0.10.11 0.10.12 0.10.13 0.10.13.0 0.11.0 0.11.1.0 0.11.1.1 0.11.1.2 0.12.1.1 0.12.1.2 0.12.1.3 0.12.1.4 0.12.1.5 0.12.1.6 0.12.1.7 0.13.2 0.13.2.1 0.13.2.3 0.90 0.90.0.1 0.90.0.3 0.90.1.0 0.90.1.1 0.91.0.0 0.92.0.0 0.92.1.0 0.92.2.0
xournal-builder: 0.1 0.1.0.0 0.1.0.1 0.1.0.2 0.1.1 0.1.1.1
xournal-convert: 0.1 0.1.0 0.1.1
xournal-parser: 0.2 0.2.0 0.3 0.3.0.0 0.3.0.1 0.3.0.2 0.4 0.4.1 0.5 0.5.0 0.5.0.1 0.5.0.2 0.5.0.3 0.5.1
xournal-render: 0.2.0.0 0.3 0.3.0.0 0.4 0.4.0.0 0.5 0.5.0 0.5.0.0 0.5.1 0.6 0.6.0
xournal-types: 0.1 0.2 0.3 0.3.0 0.3.1 0.3.2 0.4 0.5 0.5.0 0.5.0.1 0.5.0.2 0.5.1 0.5.1.0
```

--- 

As less breaking change, we could start to apply `versionNormalise` in `Cabal` whenever it reads a version from somewhere (constaints, package file)adej/5d5ac2be247fb2c81b521eef63c46402) is broken)

Yet: 41 packages with trailing zeroes used as minor or patch level bump: 
http://hdiff.luite.com/cgit/DSA/diff?id=1.0.0&id2=1.0

```
DSA: 1 1.0 1.0.0 1.0.1
Dangerous: 0.1 0.1.1 0.1.2 0.2 0.2.1 0.2.2 0.2.3 0.3 0.3.0 0.3.1 0.3.2
RSA: 1.0.0 1.0.1 1.0.2 1.0.3 1.0.4 1.0.5 1.0.6 1.0.6.1 1.0.6.2 1.0.6.3 1.2.0.0 1.2.1.0 1.2.2.0 2 2.0 2.0.0 2.1.0 2.1.0.1 2.1.0.2 2.1.0.3 2.2.0
Win32: 2.1 2.1.0.0 2.2.0.0 2.2.0.1 2.2.0.2 2.2.1.0 2.2.2.0 2.3.0.0 2.3.0.1 2.3.0.2 2.3.1.0 2.3.1.1
Workflow: 0.1 0.2 0.3 0.5.5 0.5.6 0.5.8 0.5.8.0 0.5.8.1 0.5.8.2 0.6.0.0 0.7.0.0 0.7.0.1 0.7.0.2 0.7.0.3 0.7.0.4 0.7.0.5 0.7.0.6 0.7.0.7 0.8.0.0 0.8.0.1 0.8.0.2 0.8.0.3 0.8.0.4 0.8.0.5 0.8.0.6 0.8.0.7 0.8.0.8 0.8.0.9 0.8.1 0.8.2 0.8.3
binary-derive: 0.1 0.1.0
coroutine-object: 0.1 0.1.0 0.2 0.2.0 0.2.0.0 0.3 0.3.0
data-category: 0.0.1 0.0.2 0.0.3 0.0.3.1 0.1.0 0.2.0 0.3.0 0.3.0.1 0.3.0.2 0.3.1 0.3.1.1 0.4 0.4.1 0.5.0 0.5.1 0.5.1.0 0.5.1.1 0.6.0 0.6.1
diff3: 0.1.0.0 0.1.0.1 0.1.2 0.2 0.2.0 0.2.0.1 0.2.0.2 0.2.0.3
extensible-effects: 1.0 1.1.0 1.2.0 1.2.1 1.4.1 1.5.0 1.6.0 1.7.1 1.7.1.0 1.7.1.1 1.7.1.2 1.7.2.1 1.8.0.0 1.8.1.0 1.9.0.0 1.9.0.1 1.9.1.0 1.9.2.2 1.10.0.1 1.11.0.0 1.11.0.1 1.11.0.2 1.11.0.3 1.11.0.4
fficxx: 0.1 0.1.0 0.2 0.2.1
homplexity: 0.2 0.2.0.0 0.3.0.0 0.4.0.0 0.4.1.0 0.4.2.0 0.4.3.0 0.4.3.1 0.4.3.2 0.4.3.3
hoodle: 0.1 0.1.0 0.1.0.0 0.1.0.1 0.1.1 0.1.1.0 0.1.1.1 0.2 0.2.0 0.2.1 0.2.2 0.2.2.0 0.2.2.1 0.3 0.4 0.4.0
hoodle-builder: 0.1 0.1.0 0.1.1 0.2 0.2.1 0.2.2 0.2.2.0 0.3 0.3.0
hoodle-core: 0.7 0.7.0 0.7.0.0 0.7.1 0.8 0.8.0 0.9 0.9.0 0.9.0.0 0.10 0.11 0.12 0.13 0.13.0 0.13.0.0 0.14 0.15 0.15.0
hoodle-parser: 0.1 0.1.0 0.1.1 0.2 0.2.1 0.2.2 0.3 0.3.0
hoodle-publish: 0.1 0.2 0.2.0
hoodle-render: 0.1 0.1.0 0.2 0.2.0 0.2.1 0.3 0.3.1 0.3.2 0.4 0.5 0.5.0
hoodle-types: 0.1 0.1.0 0.1.1 0.2 0.2.1 0.2.2 0.3 0.3.0
hxournal: 0.5 0.5.0.0 0.5.1 0.6 0.6.0 0.6.0.0 0.6.0.1 0.6.1 0.6.1.0 0.6.2 0.6.2.0 0.6.2.1 0.6.3 0.6.4 0.6.4.0 0.6.4.1 0.6.4.2 0.6.4.3 0.6.5 0.6.5.0 0.6.6 0.6.6.1
io-capture: 0.1 0.1.0.0 0.2 0.3 1.0.0
level-monad: 0.1 0.2 0.2.1 0.2.2 0.2.2.0 0.2.2.1 0.3 0.3.1 0.4 0.4.1
lojban: 0.0 0.0.0 0.1 0.2 0.3
lzma-conduit: 0.1.2 0.1.2.1 0.1.2.2 0.4.0 0.4.1 0.4.2 0.5.2 0.5.2.1 0.5.2.2 0.5.2.3 1.0.8 1.1.0 1.1.1 1.1.2 1.1.3 1.1.3.0
monad-parallel: 0.5 0.5.1 0.5.2 0.7 0.7.1 0.7.1.1 0.7.1.2 0.7.1.3 0.7.1.4 0.7.2 0.7.2.0 0.7.2.1 0.7.2.2
open-union: 0.1.0 0.1.0.0 0.1.0.1
pipes-cereal: 0.1.0 0.1.0.0
poppler: 0.11.0 0.11.1 0.12.0 0.12.1 0.12.2 0.12.2.1 0.12.2.2 0.12.3 0.13 0.13.1 0.14 0.14.0
scgi: 0.1 0.1.0 0.2 0.2.1 0.3 0.3.1 0.3.2 0.3.3
spelling-suggest: 0.5.0 0.5.0.1 0.5.1 0.5.1.0 0.5.2.0 0.5.2.1
storablevector: 0.1 0.1.0 0.1.1 0.1.2 0.1.2.2 0.2 0.2.1 0.2.2 0.2.3 0.2.3.1 0.2.4 0.2.5 0.2.6 0.2.7 0.2.7.1 0.2.7.2 0.2.7.3 0.2.8 0.2.8.1 0.2.8.2 0.2.8.3 0.2.9 0.2.9.1 0.2.10
stream-monad: 0.1 0.1.1 0.1.1.0 0.2 0.3 0.4 0.4.0.1 0.4.0.2
t3-game: 0.1.0 0.1.0.0 1.1.0 2.0.0
ttrie: 0.1 0.1.0.0 0.1.1 0.1.2 0.1.2.1
typography-geometry: 1.0 1.0.0 1.0.0.1
wxcore: 0.10.1 0.10.2 0.10.3 0.10.4 0.10.5 0.10.6 0.10.7 0.10.8 0.10.9 0.10.10 0.10.11 0.10.12 0.10.13 0.10.13.0 0.11.0 0.11.1.0 0.11.1.1 0.11.1.2 0.12.1.1 0.12.1.2 0.12.1.3 0.12.1.4 0.12.1.5 0.12.1.6 0.12.1.7 0.13.2 0.13.2.1 0.13.2.3 0.90 0.90.0.1 0.90.0.3 0.90.1.0 0.90.1.1 0.91.0.0 0.92.0.0 0.92.1.0 0.92.2.0
xournal-builder: 0.1 0.1.0.0 0.1.0.1 0.1.0.2 0.1.1 0.1.1.1
xournal-convert: 0.1 0.1.0 0.1.1
xournal-parser: 0.2 0.2.0 0.3 0.3.0.0 0.3.0.1 0.3.0.2 0.4 0.4.1 0.5 0.5.0 0.5.0.1 0.5.0.2 0.5.0.3 0.5.1
xournal-render: 0.2.0.0 0.3 0.3.0.0 0.4 0.4.0.0 0.5 0.5.0 0.5.0.0 0.5.1 0.6 0.6.0
xournal-types: 0.1 0.2 0.3 0.3.0 0.3.1 0.3.2 0.4 0.5 0.5.0 0.5.0.1 0.5.0.2 0.5.1 0.5.1.0
```

--- 

As less breaking change, we could start to apply `versionNormalise` in `Cabal` whenever it reads a version from somewhere (constaints, package file)

**Comment by hvr** [comment]
From @ekmett's suggestion, I think we should implement the sub-item

> make it so it doesn't accept new instances of this problem and not be stuck explaining this for decades to come.

ASAP in hackage-server, as this would stop the infection from spreading

I'm not sure we need to delete historic accidents from hackage's DB (especially since deleting artifacts from Hackage is a very messy process).

**Comment by phadej** [comment]
https://github.com/haskell/hackage-server/pull/518 the PR to `hackage-server`.

**Comment by 23Skidoo** [comment]
Closing in favour of https://github.com/haskell/hackage-server/pull/518. Feel free to reopen if you disagree.

**Pull request closed without merging**